### PR TITLE
FIX: Vue3 => Error info when creating NavLink is not shown to user

### LINKED ISSUE #11937: Error info when creating NavLink is not shown to user
**Setup**
- Rancher version: dev of 2.10.0

**Describe the bug**
- Errors when creating a NavLink are not shown to the user
- Note - this works fine for other resources (pods, services, secrets)

**To Reproduce**
- Cluster Explorer --> Header `Resource Search` --> type in NavLink --> click on NavLink --> Create
- Enter `AAA` for the name --> Enter anything for `Destination URL` --> Create

**Result**
- The Create button shows `Error`, however no other information shown on screen

**Expected Result**
In 2.9.0

image

### PR DESCRIPTION
**Pull request by aalves08**
### Summary
Fixes #11937 

### Occurred changes and/or fixed issues

- Add top-level non-custom element in order to fix error information when creating NavLink is not shown to user

### Technical notes summary

Although the changes look bigger, what was needed to make it work was just adding a `div` as the outermost container to wrap the `CruResource` component. No errors were being thrown, so I don't know the root cause for this.
Also, it may be important to cover all of the `CruResource` usages to check if it's a general problem or not and act accordingly.

### Areas or cases that should be tested

- Cluster Explorer --> Header `Resource Search` --> type in NavLink --> click on NavLink --> Create
- Enter `AAA` for the name --> Enter anything for `Destination URL` --> Create
- Should show the error from the network request to the user

### Screenshot/Video

https://github.com/user-attachments/assets/84a1d175-54e1-463e-a95c-b2b2cd514952

### Checklist
- [x] The PR is linked to an issue and the linked issue has a Milestone, or no issue is needed
- [x] The PR has a Milestone 
- [x] The PR template has been filled out
- [x] The PR has been self reviewed 
- [x] The PR has a reviewer assigned
- [x] The PR has automated tests or clear instructions for manual tests and the linked issue has appropriate QA labels, or tests are not needed
- [x] The PR has reviewed with UX and tested in light and dark mode, or there are no UX changes

### CODE REVIEW

**Comment by aalves08** [comment]
@richard-cox added you as a reviewer not only because I couldn't find the exact root cause but because of the implications this might have on other areas and potencial checks to be done. Let me know if you want to get together to look into this

**Comment by rak-phillip** [comment]
This looks like another instance where `inheritAttrs: false` would resolve the issue. I'm currently testing #11971 to see if there's a broader approach that we can take address this.

**Comment by aalves08** [comment]
> This looks like another instance where `inheritAttrs: false` would resolve the issue. I'm currently testing #11971 to see if there's a broader approach that we can take address this.

@rak-phillip I am interested in knowing more about the root cause of this.... I never understood why adding a root element would fix this 😛

**Comment by rak-phillip** [comment]
This all has to do with how Vue3 manages fallthrough attributes[^1] for root-level ancestor components - I've created a minimal repro to demonstrate the behavior:

https://play.vuejs.org/#eNqVUk1v2kAQ/SujvdBKEVbTnpCLRCkhVBQqWvW0F8ceg6nZ3e5HQmTx3zO7xjYoKCKSZXtm3uy+N/MqNlKq/+iQDVhsUl0oO+Si2CmpLYw3RZlBruUOev0oRB7a44IL3AdIhnniSgsVFwCppD6BwpoBVMfuw42vZIlNPnysUQAardOiiXxspNMpDqB3N5rN6XyfPfgPveiJo5YaBRZ3qkwsUgQQ19cMmjO+ctb8cgYRYeKobWA3zJpUirxY97dGChIdSHDmqRcl6qWyhRSGM1JQ0+AsKUv59CPkrHYYBIWeDab/LuS3Zu9znP0iJqgfiUhbs4leo63Lk98L3NN/W9zJzJWEfqO4Im2l8xxr2DcnMqJ9ggtsZ2GBhVj/MZO9RWEaUZ5omGvAc0brHL8hvaP7uf+l2QdNsfXCJdtMdSKyc+90qXcY6OSczkVwjY3aFEAuJdnqIdG944SOxrraXycsrjMZFzSgc8FnU7okXGmpvOZG2Csh9llRuHzYYmpbIRr/u0Jj1q31alWbT8P7yXy+hLvV8idMV6PF9/H9bDGNI6rUiNvh36R0CDLvCFH5NpSrqk3CIdx1qv/wAoIebws=

`App.vue` supplies `:resource` to `Child.vue; since `Child.vue` doesn't declare a `resource` prop, `resource` is applied as an attribute to the root-level element (GranChild), overwriting the `:resource="resource"` declaration. 

This can be fixed in several ways:

- Adding a div wrapper to the child component, causing the `resource` attribute to be applied to the div instead (the change in this PR)
- Applying `inheritAttrs: false` to the child component, this will signal that we don't want to automatically apply fallthrough attributes to the root-level element
- Changing the name of the conflicting attribute (#11971); this will still apply a fallthrough attribute to the root-level element, but in a way that will avoid conflicts
- Use either v-bind.attr="$data" or v-bind.prop="$data" to prevent the fallthrough behavior[^2]; this can allow us to define how we want fallthrough attributes to be applied from the parent

[^1]: https://vuejs.org/guide/components/attrs.html
[^2]: https://vuejs.org/api/built-in-directives.html#v-bind

**Comment by richard-cox** [comment]
@aalves08 can you double check if adding `inheritAttrs: false` instead of a `div` wrapper fixes the issue? If so we should wait on the conclusion of https://github.com/rancher/dashboard/pull/11971 and re-test after that merges

**Comment by aalves08** [comment]
@richard-cox I can confirm that `inheritAttrs: false` instead of a `div` does indeed fix this. Will wait for https://github.com/rancher/dashboard/pull/11971 to be merged and I'll re-test this. Thank you 🙏

**Comment by aalves08** [comment]
@richard-cox https://github.com/rancher/dashboard/pull/11971 did not fix this. Used `inheritAttrs: false` instead of a `div` and problem is fixed. Ready for re-review 🙏

**Comment by rak-phillip** [comment]
> @richard-cox #11971 did not fix this. Used `inheritAttrs: false` instead of a `div` and problem is fixed. Ready for re-review 🙏

hmm that gives me quite a bit of pause.. perhaps the solution implemented in #11971 is not the correct one for addressing this problem at the root. I'd like to investigate the issue here a little bit to better understand what else might be causing the issue.

**Comment by rak-phillip** [comment]
I've narrowed this one down to the `errors` data property from the `create-edit-view` mixin:

https://github.com/rancher/dashboard/blob/76147e00e8007cc08d8661119ef3313ef32fd744/shell/mixins/create-edit-view/impl.js#L15-L18

This is a different manifestation of the same issue, fallthrough attributes for root-level ancestor components are overwriting props. I don't intend to block the merging of this PR and I think that the `inheritAttrs: false` approach is the right one to take to solve this problem.

I'm also going to investigate alternative solutions to see if we can prevent this issue from popping up in other components.

**Comment by aalves08** [comment]
@richard-cox can this fix be approved and merged or should we wait for more feedback from @rak-phillip ?

**Comment by richard-cox** [comment]
There's something still very odd about this. If the error is down to a a mixin containing a data `error` it should also affect other similar places but it doesn't. I've made the edit experience for secrets and persistentvolumes at a high level match that of navlinks and they both correctly show errors caused by request failures.

The hierarchy is strange

- ResourceDefinition(grandparent)
  - contains CreateEditView mixin
    -  contains data: { error }
  - contains the `edit` Component (parent)
    - contains CreateEditView mixin
      -  contains data: { error }
    - contains CRUResource (child)
      - has an error prop
      - shows banner if has error

So errors is passed around in different ways. However isn't the same everywhere... in surprising ways

- ResourceDefinition - `this.errors` is always empty.
  -  Kind of expected. it's own context of CreateEditView isn't making the http request
- `edit` component - `this.errors` is always populated
  - This is good, as it passes down to CRUResource via `:errors="errors"`
- CRUResource - `this.errors` populated for services, configmaps. NOT populated for navlinks

I would expect errors to come from ResourceDefinition (via mixin) --> edit component --> cruresouce and be broken everywhere, which isn't happening.

Basically looking for some clarification / comfort that the problem here isn't universal

**Comment by rak-phillip** [comment]
@richard-cox this is explained by fallthrough attributes overwriting props in root-level ancestor components. in

In components like `shell/components/ResourceDetail/index.vue`, we bind the entire data property to dynamically rendered components

https://github.com/rancher/dashboard/blob/5405402875d336264822389a523f08f7c217a9ed/shell/components/ResourceDetail/index.vue#L422-L437

`v-bind="$data"` will attempt bind each property defined in `data` to the child component as a prop. The child component will handle and manage any matching props that are defined (in this PR, `errors`); any unmatched props in the child component will be applied to the root-level element of the child component via `$attrs`. 

`ui.cattle.io.navlink.vue` does not define an `errors` prop in the component, nor is the `errors` prop defined by any of the mixins that it consumes. Therefore, `errors` will fallthrough to the root-level component within `ui.cattle.io.navlink.vue`, overwriting `    :errors="fvUnreportedValidationErrors"` in the root-level component and causing the bug described in the issue

https://github.com/rancher/dashboard/blob/5405402875d336264822389a523f08f7c217a9ed/shell/edit/ui.cattle.io.navlink.vue#L281-L293

I think that the overall problem is two-fold:

1. A breaking change in behavior related to fallthrough attributes and root-level ancestor components - Vue2 did not overwrite props in root-level ancestors, while Vue3 does
2. Lack of a well-defined contract between `shell/components/ResourceDetail/index.vue` and the components it dynamically renders. Any given component that is dynamically rendered by ResourceDetail can experience conflict if it's not prepared to consume a data prop supplied to it by ResourceDetail. 

The way that we have been addressing this issue is to add `inheritAttrs: false` to affected components. I think that this is the best approach that we have right now.

**Review by rak-phillip — APPROVED** [review]

**Comment by richard-cox** [comment]
@rak-phillip I get the general failure, but trying to understand why it doesn't also make pages like [persistentvolumes](https://github.com/rancher/dashboard/blob/5405402875d336264822389a523f08f7c217a9ed/shell/edit/persistentvolume/index.vue#L217-L228) fail. Something to do with vue and the v-if/else?

[secrets](https://github.com/rancher/dashboard/blob/5405402875d336264822389a523f08f7c217a9ed/shell/edit/secret/index.vue#L312-L324) now seem clear (given cruresource is wrapped in a root level form element)

**Comment by rak-phillip** [comment]
@richard-cox `persistentvolumes` already has `inheritAttrs: false` specified, which would explain why we don't see the issue with `errors` there

https://github.com/rancher/dashboard/blob/5405402875d336264822389a523f08f7c217a9ed/shell/edit/persistentvolume/index.vue#L27

wrt to `secrets`, that's correct, the root-level form element will have `$attrs` applied by default.

**Comment by richard-cox** [comment]
Grand, that's all good now. Then we just need to run through the `edit` components and check where the root component is the cruresource that would receive the wrong `errors`. If there's no other we can get this in and add this to the vue3 extension docs guide. If there's lots we might want to consider a generic fix

**Comment by richard-cox** [comment]
Looks like this same issue appears in a number of different places (some aren't fixed by inheritAttrs: false,)

## Standout - These need to be confirmed or fixed in 2.10.0
- shell/edit/management.cattle.io.user.vue
- shell/edit/auth/azuread.vue
- shell/edit/auth/ldap/index.vue
- shell/edit/auth/saml.vue
- shell/edit/auth/oidc.vue
- shell/edit/auth/googleoauth.vue
- shell/edit/auth/github.vue
- shell/edit/nodeDriver.vue
- shell/edit/kontainerDriver.vue

## Debatable - These need to be confirmed but not maybe fixed in 2.10.0
- shell/edit/fleet.cattle.io.cluster.vue
- shell/edit/fleet.cattle.io.clustergroup.vue

## Others - These need to be confirmed and fixed at some point
- shell/edit/cis.cattle.io.clusterscanbenchmark.vue
- shell/edit/helm.cattle.io.projecthelmchart.vue
- shell/edit/k8s.cni.cncf.io.networkattachmentdefinition.vue
- shell/edit/management.cattle.io.projectroletemplatebinding.vue
- shell/edit/monitoring.coreos.com.route.vue
- shell/edit/constraints.gatekeeper.sh.constraint/index.vue
- shell/edit/logging-flow/index.vue
- shell/edit/logging.banzaicloud.io.output/index.vue
- shell/edit/monitoring.coreos.com.alertmanagerconfig/index.vue
- shell/edit/monitoring.coreos.com.alertmanagerconfig/receiverConfig.vue
- shell/edit/monitoring.coreos.com.prometheusrule/index.vue
- shell/edit/monitoring.coreos.com.receiver/index.vue
- shell/edit/networking.istio.io.destinationrule/index.vue

## Of Note
- shell/edit/provisioning.cattle.io.cluster/import.vue (this one matches pattern... but works without inheritAttrs)
- The hack was already put in a number of affected places by https://github.com/rancher/dashboard/pull/11803

@rak-phillip i'll leave it up to you on how to split those up, and where it leaves this PR

**Comment by rak-phillip** [comment]
@richard-cox I can work through each of the components listed and put together an issue to resolve any offenders. I don't see any reason why this PR should be blocked at this point.

**Review by richard-cox — APPROVED** [review]
